examples/gcoap/gcoap_cli.c

```diff
@@ -230,6 +235,115 @@ static ssize_t _riot_board_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, vo
     }
 }
 
+static ssize_t _riot_board_handler_blocking(coap_pkt_t *pdu, uint8_t *buf, size_t len, void *ctx)
+{
+    (void)ctx;
+    gcoap_resp_init(pdu, buf, len, COAP_CODE_CONTENT);
+    coap_opt_add_format(pdu, COAP_FORMAT_TEXT);
+    size_t resp_len = coap_opt_finish(pdu, COAP_OPT_FINISH_PAYLOAD);
+
+    xtimer_sleep(3);
+
+    /* write the RIOT board name in the response buffer */
+    if (pdu->payload_len >= strlen(RIOT_BOARD)) {
+        memcpy(pdu->payload, RIOT_BOARD, strlen(RIOT_BOARD));
+        return resp_len + strlen(RIOT_BOARD);
+    }
+    else {
+        puts("gcoap_cli: msg buffer too small");
+        return gcoap_response(pdu, buf, len, COAP_CODE_INTERNAL_SERVER_ERROR);
+    }
+}
+
+/* This is the information needed when the operation is known to take some time
+ * and a separate response is sent unconditionally. See the proxy application
+ * for an example of how to use a timer to send an empty ACK when the operation
+ * takes too long. */
+#define NONBLOCKING_READY (-1)
+static struct nonblocking_info {
+    /** A value of NONBLOCKING_READY indicates that none of this is initialized */
+    ssize_t token_length;
+    uint8_t token[GCOAP_TOKENLEN_MAX];
+    sock_udp_ep_t remote;
+    xtimer_t response_timer;
+} nonblocking = { .token_length = NONBLOCKING_READY };
+
+static void _riot_board_handler_flush(void *arg)
+{
+    struct nonblocking_info *nonblocking = (struct nonblocking_info *)arg;
+    uint8_t buf[64];
+
+    printf("Timer!\n");
+
+    coap_pkt_t pdu_backend;
+    /* Just to make copy-pasted code easier (and I trust the compiler
+     * simplifies this again) */
+    coap_pkt_t *pdu = &pdu_backend;
+
+    /* Can't do that as it won't expose the token; copy/pasting instead
+    gcoap_req_init(&pdu, &buf[0], sizeof(buf), COAP_CODE_CONTENT, NULL);
+    */
+    pdu->hdr = (coap_hdr_t *)buf;
+
+    /* generate token */
+    uint16_t msgid = 4; /* FIXME determined by random dice roll as we can't access gcoap internals */
+    ssize_t res;
+    /* Responding CON because we can -- could also respond with whatever the
+     * original request came in had we persisted that */
+    res = coap_build_hdr(pdu->hdr, COAP_TYPE_CON, &nonblocking->token[0],
+                         nonblocking->token_length, COAP_CODE_CONTENT, msgid);
+
+    coap_pkt_init(pdu, buf, sizeof(buf), res);
+
+    /* end actually gcoap_req_init */
+    coap_opt_add_format(pdu, COAP_FORMAT_TEXT);
+    size_t resp_len = coap_opt_finish(pdu, COAP_OPT_FINISH_PAYLOAD);
+    if (pdu->payload_len >= strlen(RIOT_BOARD)) {
+        memcpy(pdu->payload, RIOT_BOARD, strlen(RIOT_BOARD));
+        resp_len = resp_len + strlen(RIOT_BOARD);
+    } else {
+        puts("gcoap_cli: msg buffer too small");
+        pdu->hdr->code = COAP_CODE_INTERNAL_SERVER_ERROR;
+        /* unwinding the TEXXT option*/
+        resp_len = resp_len - 1;
```

Comment:
Hmm, this is not very intuitive .. could you also add something along the lines:
```
/* ... unwinding the TEXT option added by coap_opt_finish(..., COAP_OPT_FINISH_PAYLOAD); ... */
```

Reply:
I think it's even wrong; as opt_finish adds one byte payload marker and there's another something to unwind from the coap_opt_add_format, it's gotta be more.

The better thing to do would be to get (extract, or better add a function to expose) the length that `coap_opt_finish(pdu, COAP_OPT_FINISH_NO_PAYLOAD)` would return at the current stage of writing, and store that to roll back.